### DISCUSSION TITLE
corefine_and_compute_union() return false

### DISCUSSION THREAD

**Discussion by ankofl** [CGAL/cgal#8404] [Q&A]
### After an unsuccessful attempt to perform a Boolean union operation via C++ code, I tried to perform it via CGAL_PMP.exe , but apparently I get similar results there 
![image](https://github.com/user-attachments/assets/d5be968f-02e3-495c-ae91-8ac6a67c83d9)
### I'm trying to use this:
```c
typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Polyhedron_3<K, CGAL::Polyhedron_items_with_id_3> Mesh;

namespace PMP = CGAL::Polygon_mesh_processing;

PMP::corefine_and_compute_union(mesh1, mesh2, out)
````
### always returns false, because inside
```c
corefine_and_compute_boolean_operations(... ...){
...
return CGAL::make_array( //410  
      ob.union_is_valid(), //411 - returns false
      ob.intersection_is_valid(),  //412  
      ob.tm1_minus_tm2_is_valid(), //413  
      ob.tm2_minus_tm1_is_valid()); //414   
}
```
### Previously, I have already checked mesh1 and mesh2 in the special FixMesh.h method

```c
typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Polyhedron_3<K, CGAL::Polyhedron_items_with_id_3> Mesh;

namespace PMP = CGAL::Polygon_mesh_processing;

int FixMesh(Mesh& mesh, bool logExept = false) {
    // Сшивание граничных рёбер
    PMP::stitch_borders(mesh);

    // Исправление самопересечений
    std::vector<std::pair<Mesh::Facet_const_handle, Mesh::Facet_const_handle>> self_intersections;    

    std::cout << "e/p/i [" << mesh.edges().size() << "/" << mesh.points().size();
    PMP::self_intersections(mesh, std::back_inserter(self_intersections));
    std::cerr << "/" << self_intersections.size() << "] ";
    if (!self_intersections.empty()) {        
        try {
            if (PMP::remove_degenerate_faces(mesh)) {
                std::cout << "deg_faces: success ";
            }
            else{
                std::cout << "deg_faces: failed!" << std::endl;
                return 1;
            }

            if (PMP::remove_degenerate_edges(mesh)) {
                std::cout << "deg_edges: success ";
            }
            else {
                std::cout << "deg_edges: failed!" << std::endl;
                return 1;
            }
            if (PMP::experimental::autorefine_and_remove_self_intersections(mesh, CGAL::parameters::preserve_genus(false))) {
                std::cout << "perair: success ";
            }
            else {
                std::cout << "perair: failed!" << std::endl;
                return 1;
            }
        }
        catch (const std::exception& e) {
            std::cout << "perair: exception!" << std::endl;
            if (logExept) {
                std::cerr << e.what();
            }
            return 1;
        }        
    }

    // Исправление ориентации
    if (!PMP::is_outward_oriented(mesh)) {
        std::cerr << "Mesh is not outward oriented. Attempting to orient..." << std::endl;
        PMP::orient(mesh);
    }

    // Проверка корректности после обработки
    if (!CGAL::is_valid(mesh)) {
        std::cerr << "Mesh is not valid after processing." << std::endl;
    }

    return 0;
}
```

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): win-64
* Release or debug mode: Release and Debug
* Specific flags used (if any): CGAL_EIGEN3_ENABLED
* CGAL version: CGAL 5.5
* Boost version: last
* Other libraries versions if used (Eigen, TBB, etc.): Eigen

### At the moment, I managed to get rid of the error of performing a Boolean operation on a Mesh only by applying this method with all its methods inside for both meshes as well as for the result. This is very strong (especially the PMP::experimental::autorefine(mesh) method) performance was hit.

```cpp
#pragma once

#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Polyhedron_3.h>
#include <CGAL/Polyhedron_items_with_id_3.h>

#include <CGAL/Polygon_mesh_processing/corefinement.h>
#include <CGAL/Polygon_mesh_processing/stitch_borders.h>
#include <CGAL/Polygon_mesh_processing/orientation.h>
#include <CGAL/Polygon_mesh_processing/self_intersections.h>

#include <CGAL/Polygon_mesh_processing/repair_self_intersections.h>
#include <CGAL/Polygon_mesh_processing/stitch_borders.h>
#include <CGAL/Polygon_mesh_processing/merge_border_vertices.h>
#include <CGAL/Polygon_mesh_processing/manifoldness.h>
#include <vector>
#include <iostream>
#include <CGAL/Polygon_mesh_processing/repair.h>
#include <CGAL/Polygon_mesh_processing/polygon_mesh_to_polygon_soup.h>
#include <CGAL/Polygon_mesh_processing/polygon_soup_to_polygon_mesh.h>
#include "my_timer.h"

typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Polyhedron_3<K, CGAL::Polyhedron_items_with_id_3> Mesh;

namespace PMP = CGAL::Polygon_mesh_processing;

bool fix_mesh(Mesh& mesh) {
    std::vector<std::array<FT, 3> > points;
    std::vector<std::vector<int>> polygons;
    PMP::polygon_mesh_to_polygon_soup(mesh, points, polygons);
    PMP::repair_polygon_soup(points, polygons, CGAL::parameters::geom_traits(Array_traits()));
    if (!PMP::is_polygon_soup_a_polygon_mesh(polygons)) {
        return false;
    }
    mesh.clear();
    PMP::polygon_soup_to_polygon_mesh(points, polygons, mesh);

    PMP::stitch_borders(mesh);
    PMP::duplicate_non_manifold_vertices(mesh);
    PMP::merge_duplicated_vertices_in_boundary_cycles(mesh);
    PMP::remove_connected_components_of_negligible_size(mesh);
    PMP::remove_isolated_vertices(mesh);
    PMP::remove_degenerate_faces(mesh);
    PMP::remove_degenerate_edges(mesh);
    if (PMP::does_self_intersect(mesh)) {
        if (!PMP::experimental::remove_self_intersections(mesh)) {
            return false;
        }
    }    
    if (!PMP::is_outward_oriented(mesh)) {
        PMP::orient(mesh);        
    }
    PMP::experimental::autorefine(mesh);
    return CGAL::is_valid_polygon_mesh(mesh);
}

```

**Comment by sloriot** [comment]
As you pointed out, the union is returning `false` because there is a tangency along an edge between the two models:
![image](https://github.com/user-attachments/assets/1e7c550f-349f-4108-b2eb-b910f4fc19b1)

The input are not invalid so there is nothing to "fix". 

I don't understand why you are using the experimental `autorefine()` function, especially after calling `remove_self_intersections()`. You should be calling the new function (not in experimental namespace) available in CGAL 6.0 beta 1:
https://cgal.geometryfactory.com/CGAL/doc/master/Polygon_mesh_processing/group__PMP__corefinement__grp.html#ga3e3a0a82b6c04bdc3a6c070e8da4aed5

Or even the one operating on triangle soups:
https://cgal.geometryfactory.com/CGAL/doc/master/Polygon_mesh_processing/group__PMP__corefinement__grp.html#gaec85370aa0b2acc0919e5f8406cfb74c

**Reply by ankofl** [reply]
You have no idea how much I've been waiting for your answer! Thank you so much for it!

In addition, I want to ask how it is possible to combine grids that do not intersect, but stand close to each other? I tried to combine the grids, and if the operation failed, then decompose the second of the two grids into convex components, expand each of them a little (something like the Minkowski sum) and try to connect each decomposed and enlarged part with the first grid, but this created more problems than benefits (which is probably not surprising)

So, if you have a lot of figures that stand close to each other, what would be an effective way to combine them?

**Reply by sloriot** [reply]
What do you mean with `grids`?

**Reply by ankofl** [reply]
Oh, I'm sorry, I don't speak English well and I communicate here through a translator, he mistakenly translated Mesh into Grid, I meant Mesh
```c
typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Polyhedron_3<K, CGAL::Polyhedron_items_with_id_3> Mesh;
```

In general, at the moment I am trying to transform Mesh into Nef 
```c
typedef CGAL::Exact_predicates_exact_constructions_kernel K_e;
typedef CGAL::Nef_polyhedron_3<K_e, CGAL::SNC_indexed_items> Nef;
```
and perform a Boolean operation with it, and then get Mesh from Nef, but so far too often Nef does not fulfill the condition after a Boolean operation .is_simple()
```c
Nef pre = mass.difference(constr).regularization();
if (pre.is_simple()) {
}
```
I will try to send here tomorrow a minimally feasible example with the data that I use

Thanks for the help!

**Comment by ankofl** [comment]
I get ```false``` quite often for the reason you described earlier, but I don't understand how to solve it. I have prepared a data set here, where the difference between m1 and m2 of each pair of Polyhedron_3 was performed (the code in front indicates the pair identifier)

![image](https://github.com/user-attachments/assets/c1a2e4ea-a496-4a0b-8c70-4a974b7115dc)

[failed differences.zip](https://github.com/user-attachments/files/16628265/failed.differences.zip)

**Comment by ankofl** [comment]
The problem was solved by using union rather than difference, as well as using bounding_box for some complex Meshes

**Comment by ankofl** [comment]
UPD: 14.10  @sloriot @MaelRL 

Although we previously managed to find a workaround for this problem by increasing one of the Mesh in all directions, I think it is not optimal, so I think we should continue the discussion here on what to do when 
```cpp
PMP::corefine_and_compute_union()
```
it returns false

`/h/typedefs/typedefs_mesh.h`
```cpp
#include <CGAL/Polygon_mesh_processing/polygon_soup_to_polygon_mesh.h>
#include <CGAL/IO/read_off_points.h>
#include <CGAL/Polyhedron_3.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/Polygon_mesh_processing/triangulate_faces.h>
#include <CGAL/Polygon_mesh_processing/self_intersections.h>
#include <CGAL/Polygon_mesh_processing/orientation.h>
#include <CGAL/Polyhedron_items_with_id_3.h>
#include <CGAL/Nef_polyhedron_3.h>
#include <CGAL/Nef_3/SNC_indexed_items.h>

typedef CGAL::Exact_predicates_exact_constructions_kernel K;
typedef K::Point_3                                P;
typedef K::Vector_3                               V;
typedef std::vector<std::size_t> Face;

typedef CGAL::Polyhedron_3<K> Mesh;

typedef CGAL::Surface_mesh<P> Surface;
typedef Surface::Halfedge_index Halfedge_index;
typedef Surface::Vertex_index Vertex_index;

typedef CGAL::Nef_polyhedron_3<K, CGAL::SNC_indexed_items> Nef;

namespace PMP = CGAL::Polygon_mesh_processing;

struct VectorHash {
    template <class T>
    std::size_t operator()(const T& v) const {
        // Èñïîëüçóåì õýø-ôóíêöèþ äëÿ x, y è z
        std::size_t seed = 0;
        seed ^= std::hash<double>()(CGAL::to_double(v.x())) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
        seed ^= std::hash<double>()(CGAL::to_double(v.y())) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
        seed ^= std::hash<double>()(CGAL::to_double(v.z())) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
        return seed;
    }
};

struct VectorEqual {
    bool operator()(const V& lhs, const V& rhs) const {
        // Îïðåäåëèòå ñðàâíåíèå äëÿ íîðìàëåé
        return lhs == rhs; // èëè èñïîëüçóéòå ñðàâíåíèå ñ îïðåäåëåííîé òî÷íîñòüþ
    }
};

struct PointEqual {
    bool operator()(const P& lhs, const P& rhs) const {
        // Îïðåäåëèòå ñðàâíåíèå äëÿ íîðìàëåé
        return lhs == rhs; // èëè èñïîëüçóéòå ñðàâíåíèå ñ îïðåäåëåííîé òî÷íîñòüþ
    }
};

inline bool load_from(Mesh& output, const std::string& path) {
    output.clear();
    std::ifstream input;
    input.open(path);
    if (!input) {
        return false;
    }
    else{
        input >> output;
    }
    input.close();

    if (!CGAL::is_triangle_mesh(output)) {
        if (!PMP::triangulate_faces(output)) {
            std::cout << "no_triangulate\n";
        }
    }

    if (!output.is_closed()) {
        std::cout << "no_closed\n";
    }

    if (!PMP::does_bound_a_volume(output)) {
        std::cout << "no_bound\n";
    }

    if (PMP::does_self_intersect(output)) {
        std::cout << "self_inter\n";
    }

    return CGAL::is_valid_polygon_mesh(output);
}

inline bool save_to(const Mesh& input, const std::string path) {
    if (!CGAL::is_valid_polygon_mesh(input)) {
        return false;
    }
    if (CGAL::IO::write_polygon_mesh(path, input, CGAL::parameters::stream_precision(17))) {
        return true;
    }
    return false;
}
```

[Archive with original off-files needed to union](https://github.com/user-attachments/files/17367133/Unite.zip)

[off-file of Polyhedron_3 where need remove degenerated faces](https://github.com/user-attachments/files/17367087/united.zip)

[Project, where im work](https://github.com/ankofl/Native)

```cpp
inline void union_meshes(
	const std::string& path_meshes,
	const std::string& path_mesh_output) {

	std::vector<Mesh> meshes;
	if (!load_dir(path_meshes, meshes)) {
		return;
	}

	Mesh first = meshes[0];
	for (size_t i = 1; i < meshes.size(); i++) {
		Mesh result;
		if (PMP::corefine_and_compute_union(first, meshes[i], result)) {
			first = result;
		}
		else {
		       // problem here
		}
		std::cout << i + 1 << '/' << meshes.size() << std::endl;
	}

	remesh_planar(first, first);

	save_to(first, path_mesh_output);
}
```

```cpp
inline bool load_dir(const std::string& directory, std::vector<Mesh>& meshes) {
	for (const auto& entry : std::filesystem::directory_iterator(directory)) {
		if (entry.is_regular_file() && entry.path().extension() == ".off") {
			Mesh loaded;
			if (load_from(loaded, entry.path().string())) {
				meshes.push_back(loaded);
			}
		}
	}
	return meshes.size() > 0;
}
```

**Comment by ankofl** [comment]
[A similar problem](https://github.com/CGAL/cgal/issues/2861)

I get a lot of similar cases (44 union-operations out of 1718 fail) 

As you can see from the screenshots below, the problem lies in the edges that lie on top of each other:
![Снимок экрана 2024-10-15 171017](https://github.com/user-attachments/assets/9d3e3b93-0937-4f56-bdd9-15a1195f05e7)
![Снимок экрана 2024-10-15 170739](https://github.com/user-attachments/assets/b738d0e9-2066-4bf8-a46f-dce35c2141a0)
![Снимок экрана 2024-10-15 170808](https://github.com/user-attachments/assets/191b3d9d-47ec-4179-bd6e-41e33f72ecc4)
![Снимок экрана 2024-10-15 170942](https://github.com/user-attachments/assets/10a24d75-ace5-4218-8f8a-387cafa8a61b)

Based on this, the question arises - is it technically possible for Polyhedron_3 to combine such meshes, or is expansion in all directions generally the only option to combine them?

Thanks @sloriot for the help!
